modules/internal/ChapelArray.chpl

```diff
@@ -309,33 +305,74 @@ module ChapelArray {
                       " If so, use {...} instead of [...].");
     }
 
+    param homog = isHomogeneousTuple(elems);
+
     // elements of string literals are assumed to be of type string
-    type eltType = _getLiteralType(elems(0).type);
+    type eltType = if homog then elems(0).type
+                            else chpl_computeUnifiedType(elems);
+
     var dom = {arrayLiteralLowBound..#k};
     var arr = dom.buildArray(eltType, initElts=false);
 
-    for param i in 0..k-1 {
-      type currType = _getLiteralType(elems(i).type);
-
-      if currType != eltType {
-        compilerError( "Array literal element " + i:string +
-                       " expected to be of type " + eltType:string +
-                       " but is of type " + currType:string );
+    if homog {
+      for i in 0..<k {
+        ref dst = arr(i+arrayLiteralLowBound);
+        ref src = elems(i);
+        __primitive("=", dst, src);
+      }
+    } else {
+      for param i in 0..k-1 {
+        type currType = elems(i).type;
+
+        ref dst = arr(i+arrayLiteralLowBound);
+        ref src = elems(i);
+        if (currType == eltType ||
+            Reflection.canResolve("=", dst, src)) {
+          __primitive("=", dst, src);
+        } else {
+          compilerError( "Array literal element " + i:string +
+                         " expected to be of type " + eltType:string +
+                         " but is of type " + currType:string );
```

Comment:
OK to keep this, however this should not happen, right? given that computeUnifyingType() completed successfully.  Perhaps we do not need the canResolve() check above at all?

Reply:
I agree that it shouldn't happen, but was worried about users getting a worse error message if anything went wrong down the road (say, return type unification got smarter faster than `=` between the types did).  So I view this as an assertion that has a nicer user-facing form rather than an error we anticipate getting triggered.